docker_ci_deploy/__main__.py

```diff
@@ -9,15 +9,34 @@
 from itertools import chain
 
 
+# This is complicated but these are the complete regexes used in Docker to
+# match image tags. We only use these 2 regexes as porting all of the machinery
+# from golang to Python is too much work.
+#
+# Source code in Docker:
+# https://github.com/docker/distribution/blob/v2.6.0-rc.2/reference/regexp.go
+#
+# The pattern strings were extracted using The Go Playground:
+# https://play.golang.org/p/xYRMnoqMqk
+REFERENCE_REGEX = re.compile(
+    r'^((?:(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])(?:(?:\.(?:[a-zA'
+    r'-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+)?(?::[0-9]+)?/)?[a-z0-9]+('
+    r'?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)'
+    r'[a-z0-9]+)+)?)+)?)(?::([\w][\w.-]{0,127}))?(?:@([A-Za-z][A-Za-z0-9]*(?:['
+    r'-_+.][A-Za-z][A-Za-z0-9]*)*[:][[:xdigit:]]{32,}))?$')
+
+ANCHORED_NAME_REGEX = re.compile(
+    r'^(?:((?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])(?:(?:\.(?:[a-zA'
+    r'-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+)?(?::[0-9]+)?)/)?([a-z0-9]'
+    r'+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]'
+    r'*)[a-z0-9]+)+)?)+)?)$')
```

Comment:
I'd be happier if these were less opaque. :-/

We can probably do a subset of what the docker code does. At the very least, we should factor out `NameRegexp`, `TagRegexp`, and `DigestRegexp` to reduce duplication and make the whole thing parseable by humans.